visualization/app/codeCharta/ui/codeMap/rendering/geometryGenerationHelper.ts

```diff
@@ -0,0 +1,369 @@
+import { Node } from "../../../codeCharta.model"
+import { CodeMapBuilding } from "./codeMapBuilding"
+import { Box3, Vector3, Vector2 } from "three"
+import { CodeMapGeometricDescription } from "./codeMapGeometricDescription"
+import { ColorConverter } from "../../../util/color/colorConverter"
+
+export interface BoxMeasures {
+	x: number
+	y: number
+	z: number
+	width: number
+	height: number
+	depth: number
+}
+
+export interface SurfaceInformation {
+	node: Node
+	surfaceStartIndex: number
+	minPos: Vector2
+	maxPos: Vector2
+}
+
+export interface IntermediateVertexData {
+	positions: Float32Array
+	normals: Float32Array
+	uvs: Float32Array
+	colors: Float32Array
+
+	indices: Uint32Array
+	ids: Float32Array
+
+	deltas: Float32Array
+	isHeight: Float32Array
+
+	floorSurfaceInformation: SurfaceInformation[]
+}
+
+enum sides {
+	left = 0,
+	right = 1,
+	bottom = 2,
+	top = 3,
+	back = 4,
+	front = 5
+}
+
+const normals: Vector3[] = [
+	new Vector3(-1, 0, 0),
+	new Vector3(1, 0, 0),
+	new Vector3(0, -1, 0),
+	new Vector3(0, 1, 0),
+	new Vector3(0, 0, -1),
+	new Vector3(0, 0, 1)
+]
+
+const numberSides = 6
+const verticesPerSide = 4
+const threeDimensions = 3
+const twoDimensions = 2
+
+// Helper functions
+
+export function addBoxToVertexData(
+	data: IntermediateVertexData,
+	node: Node,
+	measures: BoxMeasures,
+	color: string,
+	subGeomIndex: number,
+	desc: CodeMapGeometricDescription,
+	delta: number,
+	addingFloor: boolean
+) {
+	desc.add(
+		new CodeMapBuilding(
+			subGeomIndex,
+			new Box3(
+				new Vector3(measures.x, measures.y, measures.z),
+				new Vector3(measures.x + measures.width, measures.y + measures.height, measures.z + measures.depth)
+			),
+			node,
+			color
+		)
+	)
+
+	// This call has side effects on `positions` and `uvs`..
+	// Each `position` entry consists out of a vector 3.
+	setPositions(data.positions, measures, subGeomIndex)
+	setUVs(data.uvs, subGeomIndex)
+	setVerticesAndFaces(node, measures, color, delta, subGeomIndex, data, addingFloor)
+}
+
+function setPositions(positions: Float32Array, measures: BoxMeasures, index: number) {
+	const { x: minPosX, y: minPosY, z: minPosZ, width, height, depth } = measures
+	const maxPosX = minPosX + width
+	const maxPosY = minPosY + height
+	const maxPosZ = minPosZ + depth
+
+	let positionIndex = index * verticesPerSide * numberSides * threeDimensions
+
+	// left // 0
+	positions[positionIndex++] = minPosX
+	positions[positionIndex++] = minPosY
+	positions[positionIndex++] = minPosZ
+
+	positions[positionIndex++] = minPosX
+	positions[positionIndex++] = maxPosY
+	positions[positionIndex++] = minPosZ
+
+	positions[positionIndex++] = minPosX
+	positions[positionIndex++] = maxPosY
+	positions[positionIndex++] = maxPosZ
+
+	positions[positionIndex++] = minPosX
+	positions[positionIndex++] = minPosY
+	positions[positionIndex++] = maxPosZ
```

Comment:
Note: I decided against using `.set()` here to prevent the required array allocation.